### PR TITLE
FC-73: Add XqueueViewset with xwatcher services

### PR DESCRIPTION
**Pull request by UsamaSadiq**
## Description
- This PR is taking on the work already done in the PR https://github.com/openedx/edx-submissions/pull/287. 

----
# FC-73 Feature: XQueueViewSet - Compatible API Endpoints for External Grader Integration

⚠️ **Important:** This PR builds on the `ExternalGraderDetail` and `SubmissionFile` infrastructure deployed in previous PRs (FC-73). Please ensure those changes are fully deployed before merging.

## Description

This pull request implements the XQueueViewSet, providing compatible API endpoints that allow external graders (xqueue-watcher) to interact with the new submission and grading architecture. This viewset consolidates authentication, submission retrieval, and result processing services while leveraging the new `ExternalGraderDetail` and `SubmissionFile` models.

## Motivation

The current XQueue implementation requires a strategic update to:
* Provide a secure and efficient API for external grader integration
* Simplify the complex multi-system communication patterns
* Enhance session management and security
* Integrate cleanly with the new submission architecture

### Authentication Services
* Custom `XQueueSessionAuthentication` class with:
  * CSRF exemptions for result endpoints
  * Secure login/logout endpoints

### Submission Distribution
* Queue-based submission retrieval through `get_submission` endpoint
* Integrated file handling through `SubmissionFileManager`
* Status tracking with explicit state transitions
* UUID-based submission keys for enhanced security

### Result Processing
* Transactional score updates via the `put_result` endpoint
* Comprehensive error handling with detailed logging
* Atomic status updates for reliable state management
* Automatic retry mechanism for failed submissions

### Score Processing
* Integration with the submissions API via `set_score`
* Robust failure tracking and processing
* Status transition enforcement

### Error Handling
* Comprehensive validation of grader responses
* Detailed logging for troubleshooting
* Clear error responses for client integration

## Testing Strategy

Comprehensive test coverage includes:
* Authentication workflow verification
* Submission retrieval process
* Result processing and error handling
* Session management edge cases
* Integration with existing submission models

## Open edX Compliance

This implementation adheres to Open edX standards through:
* RESTful API design following DRF best practices
* Comprehensive security measures
* Transaction safety for critical operations
* Extensive logging for operational visibility
* Compatibility with existing systems

## Performance Considerations

* Optimized database queries with select_related
* Efficient transaction handling
* Minimal processing overhead

**BREAKING CHANGES**: None. Designed for full backward compatibility with existing xqueue-watcher services.

## Documentation

Updated documentation will include:
* XQueue API specification for external graders
* Session management guidelines
* Error handling and response format specifications
* Integration guide for external grader services

## Implementation References

* Related ADR: Implementation of XQueue Compatible Views for External Grader Integration
* Previous PRs: FC-73 SubmissionQueueRecord and SubmissionFile implementations

### CODE REVIEW

**Comment by UsamaSadiq** [comment]
After reviewing the existing implementation, the only point that came up was to secure all XQueueViewset endpoints with Grader-level authentication. Once that is implemented, the PR will be ready for review.

File: submissions/urls.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+from rest_framework.routers import DefaultRouter
+
+from submissions.views.xqueue import XQueueViewSet
+
+router = DefaultRouter()
+router.register(r'', XQueueViewSet, basename='xqueue')
+
+urlpatterns = [
+    path('', include(router.urls)),
+]
```
Comment (ormsbee) [inline submissions/urls.py:15]: Question: Does it have to be at the root like this in order to be backwards compatible with xqueue-watcher expectations?

File: submissions/permissions.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+from rest_framework import permissions
+
+
+class IsXQueueUser(permissions.BasePermission):
+    """
+    Permission classes for submissions app.
+    """
+
+    def has_permission(self, request, view):
+        return request.user.groups.filter(name='xqueue').exists()
```
Comment (ormsbee) [inline submissions/permissions.py:11]: It doesn't have to happen in this PR, but this expectation is going to need to go in the README for edx-submissions and in the release notes.

**Review by ormsbee — APPROVED** [review]
I have a minor question and note, but nothing blocking. Please squash and I'll merge on Monday morning.

**Comment by UsamaSadiq** [comment]
@ormsbee I've updated the package version. Should I go ahead to merge and tag the new release or would you prefer to do it yourself?
